In [24]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [25]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [26]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [27]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [28]:
df_market_data.dtypes

price_change_percentage_24h     float64
price_change_percentage_7d      float64
price_change_percentage_14d     float64
price_change_percentage_30d     float64
price_change_percentage_60d     float64
price_change_percentage_200d    float64
price_change_percentage_1y      float64
dtype: object

In [29]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d",
                                                              "price_change_percentage_30d", "price_change_percentage_60d", "price_change_percentage_200d",
                                                              "price_change_percentage_1y"]])


In [35]:
# Create a DataFrame with the scaled data
df_crypto_scaled = pd.DataFrame(crypto_scaled, columns=["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d",
                                                              "price_change_percentage_30d", "price_change_percentage_60d", "price_change_percentage_200d",
                                                              "price_change_percentage_1y"])

# Copy the crypto names from the original data
df_market_data.reset_index(inplace=True)
df_crypto_scaled['coinID']= df_market_data['coin_id']
# Set the coinid column as index
df_crypto_scaled=df_crypto_scaled.set_index('coinID')

In [31]:
# Display sample data
df_crypto_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
0                     0.508529                    0.493193   
1                     0.185446                    0.934445   
2                     0.021774                   -0.706337   
3                    -0.040764                   -0.810928   
4                     1.193036                    2.000959   

   price_change_percentage_14d  price_change_percentage_30d  \
0                     0.772200                     0.235460   
1                     0.558692                    -0.054341   
2                    -0.021680                    -0.061030   
3                     0.249458                    -0.050388   
4                     1.760610                     0.545842   

   price_change_percentage_60d  price_change_percentage_200d  \
0                    -0.067495                     -0.355953   
1                    -0.273483                     -0.115759   
2                     0.008005                     -0.550247   
3                    -0.373164                     -0.458259   
4                    -0.291203                     -0.499848   

   price_change_percentage_1y        coinID  
0                   -0.251637       bitcoin  
1                   -0.199352      ethereum  
2                   -0.282061        tether  
3                   -0.295546        ripple  
4                   -0.270317  bitcoin-cash

In [36]:
df_crypto_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coinID                                                                  
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coinID                                                                   
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coinID                                                                    
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coinID                                    
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [37]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [40]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1)
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(df_crypto_scaled)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)

C:\Users\Pooja\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [41]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

In [42]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [43]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [44]:
# Fit the K-Means model using the scaled data
model.fit(df_crypto_scaled)

KMeans(n_clusters=4, random_state=1)

In [45]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_predict = model.predict(df_crypto_scaled)

# Print the resulting array of cluster values.
print(k_predict)

[3 3 1 1 3 3 3 3 3 1 1 1 1 3 1 3 1 1 3 1 1 3 1 1 1 1 1 1 3 1 1 1 0 3 1 1 2
 1 1 1 1]


In [46]:
# Create a copy of the DataFrame
crypto_scaled_copy = df_crypto_scaled.copy(deep = True)

In [48]:
# Add a new column to the DataFrame with the predicted clusters
df_crypto_scaled["cluster_predict"] = k_predict

# Display sample data
df_crypto_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coinID                                                                  
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coinID                                                                   
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coinID                                                                    
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster_predict  
coinID                                                     
bitcoin                        -0.251637                3  
ethereum                       -0.199352                3  
tether                         -0.282061                1  
ripple                         -0.295546                1  
bitcoin-cash                   -0.270317                3

In [51]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
df_crypto_scaled.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    hover_cols="all"
).opts(yformatter="%.0f")

:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coinID,price_change_percentage_14d,price_change_percentage_30d,price_change_percentage_60d,price_change_percentage_200d,price_change_percentage_1y,cluster_predict)

---

### Optimize Clusters with Principal Component Analysis.

In [52]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components = 3)

In [53]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
crypto_pca = pca.fit_transform(df_crypto_scaled)
# View the first five rows of the DataFrame. 
crypto_pca[:5]

array([[-1.58513942,  0.13439444,  0.545621  ],
       [-1.32258066,  0.05449393,  1.11570854],
       [ 0.26633536, -0.53640937, -0.79694588],
       [ 0.25722513, -0.59201286, -0.6465287 ],
       [-2.95224985,  0.42237155,  1.26789353]])

In [55]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
explained_var = pca.explained_variance_ratio_
explained_var

array([0.37005408, 0.32322221, 0.19115222])

In [56]:
explained_var.sum()

0.8844285111826466

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8844285111826466

In [115]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
crypto_pca_df = pd.DataFrame(crypto_pca)

# Copy the crypto names from the original data
crypto_pca_df['coinID'] = df_market_data['coin_id']

# Set the coinid column as index
crypto_pca_df = crypto_pca_df.set_index('coinID')

crypto_pca_df.rename(columns={0: 'PC1',
                    1: 'PC2',
                    2: 'PC3'},
          inplace=True, errors='raise')

# Display sample data
crypto_pca_df.head()



PC1       PC2       PC3
coinID                                    
bitcoin      -1.585139  0.134394  0.545621
ethereum     -1.322581  0.054494  1.115709
tether        0.266335 -0.536409 -0.796946
ripple        0.257225 -0.592013 -0.646529
bitcoin-cash -2.952250  0.422372  1.267894

---

### Find the Best Value for k Using the PCA Data

In [116]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1,11))

In [117]:
# Create an empy list to store the inertia values
inertia2 = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1)
# 2. Fit the model to the data using `df_market_data_pca`
    k_model.fit(crypto_pca_df)
# 3. Append the model.inertia_ to the inertia list
    inertia2.append(k_model.inertia_)


C:\Users\Pooja\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [118]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_2 = {"k": k_values, "inertia": inertia2}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow2 = pd.DataFrame(elbow_data_2)
df_elbow2.head()

k     inertia
0  1  287.007838
1  2  196.268358
2  3  111.972027
3  4   43.205879
4  5   33.042435

In [119]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow2.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No; it is the same.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [120]:
# Initialize the K-Means model using the best value for k
model2 = KMeans(n_clusters=4, random_state=0)

In [121]:
# Fit the K-Means model using the PCA data
model2.fit(crypto_pca_df)

KMeans(n_clusters=4, random_state=0)

In [122]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_3 = model2.predict(crypto_pca_df)
# Print the resulting array of cluster values.
print(k_3)

[0 0 2 2 0 0 0 0 0 2 2 2 2 0 2 0 2 2 0 2 2 0 2 2 2 2 2 2 0 2 2 2 1 0 2 2 3
 2 2 2 2]


In [123]:
# Create a copy of the DataFrame with the PCA data
crypto_pca_df_copy = crypto_pca_df.copy(deep = True)

# Add a new column to the DataFrame with the predicted clusters
crypto_pca_df['Predicted_Clusters'] = k_3

# Display sample data
crypto_pca_df.head()
crypto_pca_df.columns

Index(['PC1', 'PC2', 'PC3', 'Predicted_Clusters'], dtype='object')

In [124]:
crypto_pca_df.head()

PC1       PC2       PC3  Predicted_Clusters
coinID                                                        
bitcoin      -1.585139  0.134394  0.545621                   0
ethereum     -1.322581  0.054494  1.115709                   0
tether        0.266335 -0.536409 -0.796946                   2
ripple        0.257225 -0.592013 -0.646529                   2
bitcoin-cash -2.952250  0.422372  1.267894                   0

In [125]:
crypto_pca_df.hvplot.scatter(
    x="PC1",
    y="PC2",
    z="PC3",
    hover_cols="all"
).opts(yformatter="%.0f")

:Scatter   [PC1]   (PC2,coinID,PC3,Predicted_Clusters)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [93]:
pca_plot = df_elbow2.hvplot.line(
    x="k", 
    y="inertia", 
    title="Crypto Elbow Curve", 
    xticks=k
)

non_pca_plot = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="PCA Elbow Curve", 
    xticks=k
)

In [94]:
# Composite plot to contrast the Elbow curves

(pca_plot +non_pca_plot).cols(1)

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [129]:
# Composite plot to contrast the clusters
pca_scatter = crypto_pca_df.hvplot.scatter(
    x="PC1",
    y="PC2",
    hover_cols="all"
).opts(yformatter="%.0f")

initial_scatter =df_crypto_scaled.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    hover_cols="all"
).opts(yformatter="%.0f")

# Composite Graph
(pca_scatter +initial_scatter).cols(1)

:Layout
   .Scatter.I  :Scatter   [PC1]   (PC2,coinID,PC3,Predicted_Clusters)
   .Scatter.II :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coinID,price_change_percentage_14d,price_change_percentage_30d,price_change_percentage_60d,price_change_percentage_200d,price_change_percentage_1y,cluster_predict)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The clusters appear more concentrated (as it has a reduced variance).